### DISCUSSION TITLE
How to ensure a component remains at the bottom layer of a graph

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#1135] [Q&A]
Hello,
I am currently working on a graph and I need to ensure that a specific component is always positioned underneath other component types. I am looking for a way to implement this.

In my current scenario, I have created a rectangular shape that is positioned at the bottom layer of the graph. However, when I create other components before the rectangular shape, it tends to move up and hide other components. 

thanks in advance.

![image](https://github.com/eclipse-glsp/glsp/assets/116633370/9c2c359a-8605-4b5e-bfbc-d6351edb5b12)

**Comment by planger** [comment]
Hi @Adem-Kri,

currently, the only way to control the layering (z axis) is to maintain a corresponding order in the graph model. The graph model is rendered in a top-town child first traversal, which results in their respective SVG elements to be put in the order of the traversal. According to the painters model of SVG, the element that is rendered last, will be on top of the elements that have been rendered before.

Thus, you need to make sure that your graph model has the elements that should be rendered on top further down the element list. If they are in the same parent element, this can be done by changing the order of the parent element's children. If they are in different parent elements, you need to make sure that the parent of the element that needs to go on top is after the parent element that needs to go to the back.

We are considering introducing an explicit layering mechanism in the graph model, see also https://github.com/eclipse-glsp/glsp/issues/501. However, we didn't get to this issue yet.
It is also worth mentioning that there is a `BringToFrontAction` / [`BringToFrontCommand`](https://github.com/eclipse-sprotty/sprotty/blob/master/packages/sprotty/src/features/zorder/zorder.ts). However, this only moves the element within the same parent to the front.
